Workflow Editor Activity Bar

ElectronicBlueberry opened the pull request:
This PR brings the activity bar to the workflow editor, providing various workflow-related activities and tools. Key changes and features include:

- Support for undo/redo stack UI.
- Add workflows activity item to explore in detail and select workflows to add/copy steps from.
- Redesign expandable tool attributes panel
- More logic and UI improvements

<img width="1512" alt="Screenshot 2024-11-15 at 20 59 33" src="https://github.com/user-attachments/assets/0865aea5-9728-441e-9628-c529184b0d7f">

https://github.com/user-attachments/assets/c0c9301c-d8f7-44a2-a2f3-7e77b73a23cf

Chnages

- changes (undo redo stack ui)
- workflows (to insert like tools)

Ideas for additional activities, but unsure about:

- workflow specific utility tools
- workflow attributes
  this would free up the right panel to serve a single purpose
- step list
- subworkflows (in this workflow)
- invocations of this workflow 
- run with simplified run form
- edit report 

## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [x] Instructions for manual testing are as follows:
  1. Open the worklfow editore and play around

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

ElectronicBlueberry committed:
generalize activity bar

ElectronicBlueberry committed:
add activity bar to workflow editor

ElectronicBlueberry committed:
add activities to activity bar
debounce sync to avoid staggered calls
pass activity bar id where needed

ElectronicBlueberry committed:
add save and exit activity

ElectronicBlueberry committed:
route to workflow list on exit

ElectronicBlueberry committed:
move best practices to activity bar

ElectronicBlueberry committed:
use ActivityPanel component for Lint

ElectronicBlueberry committed:
move UndoRedoStack to activity bar

ElectronicBlueberry committed:
add placeholder activities

ElectronicBlueberry committed:
fix delayed input
make delayed input v-model compliant

ElectronicBlueberry committed:
make favorites button reusable

ElectronicBlueberry committed:
refactor WorkflowCard
add types
fix type errors
remove redundant code
remove any
unify naming

ElectronicBlueberry committed:
improve performance of workflow card
make workflow card mostly stateless
move modals outside of workflow card

ElectronicBlueberry committed:
add workflows panel

ElectronicBlueberry committed:
fix z-indexing for dropdown
add missing actions to dropdown

ElectronicBlueberry committed:
connect insert and insert steps events

ElectronicBlueberry committed:
move breakpoints to partial
remove unused css
fix broken css
add non-found indicator

ElectronicBlueberry committed:
only occupy vertical space if invocations count is present

ElectronicBlueberry committed:
rename Attributes to WorkflowAttributes

ElectronicBlueberry committed:
make attributes an activity

ElectronicBlueberry committed:
move MarkdownEditor to activity bar

ElectronicBlueberry committed:
move all options to activity bar

ElectronicBlueberry committed:
add save button

ElectronicBlueberry committed:
remove unused attribute and unreachable code

ElectronicBlueberry committed:
enable run and save

ElectronicBlueberry committed:
fix activity text centering on chrome

ElectronicBlueberry committed:
fix text short property collision

ElectronicBlueberry committed:
decouple node inspector from right hand panel

ElectronicBlueberry committed:
modularize drag handle

ElectronicBlueberry committed:
correct appearance of drag handle

ElectronicBlueberry committed:
show collapse button

ElectronicBlueberry committed:
add tool form to node inspector

ElectronicBlueberry committed:
add inspector heading

ElectronicBlueberry committed:
allow for saving default sizes for tools / steps

ElectronicBlueberry committed:
make double click maximize node inspector

ElectronicBlueberry committed:
fix scroll height

ElectronicBlueberry committed:
add form default to inspector

ElectronicBlueberry committed:
make form default full width

ElectronicBlueberry committed:
remove ability to sub-workflow own workflow

ElectronicBlueberry committed:
connect refresh list signal

ElectronicBlueberry committed:
fix workflow creation

ElectronicBlueberry committed:
make current workflow immutable from card

ElectronicBlueberry committed:
replace save with save and exit

ElectronicBlueberry committed:
add save button to title bar

ElectronicBlueberry committed:
move initial position to not interfere with toolbar

ElectronicBlueberry committed:
make step annotation resizable

ElectronicBlueberry committed:
fix activity item test

ElectronicBlueberry committed:
fix workflow attributes test

ElectronicBlueberry committed:
fxi activity store test

ElectronicBlueberry committed:
fix empty divs rendered

ElectronicBlueberry committed:
fix activity settings test

ElectronicBlueberry committed:
fix lint test

ElectronicBlueberry committed:
fix filter conversion test

ElectronicBlueberry committed:
fix workflow filters import

ElectronicBlueberry committed:
remove broken test case

ElectronicBlueberry committed:
remove workflows section in tool box

ElectronicBlueberry committed:
add back activity item id

ElectronicBlueberry committed:
make workflow editor default to attributes

ElectronicBlueberry committed:
add class for editor save button

ElectronicBlueberry committed:
move selector to id

ElectronicBlueberry committed:
fix auto layout

ElectronicBlueberry committed:
add button titles to node inspector

ElectronicBlueberry committed:
fix maximize center pane function

ElectronicBlueberry committed:
make auto layout respect grid and snapping

ElectronicBlueberry committed:
make auto layout an action

ElectronicBlueberry committed:
make activity panel scrollable

ElectronicBlueberry committed:
close inspector instead of showing attributes

ElectronicBlueberry committed:
open nodes from lint panel

ElectronicBlueberry committed:
auto layout frame comments

ElectronicBlueberry committed:
undo comment auto layout

ElectronicBlueberry committed:
consider comments in layout

ElectronicBlueberry committed:
move freehand comments with their closest comment preserving large drawings

ElectronicBlueberry committed:
fix edge crossing

ElectronicBlueberry committed:
associate comments to their closest node

ElectronicBlueberry committed:
make auto layout a toolbar tool

ElectronicBlueberry committed:
pass step position for more accurate comment positioning

ElectronicBlueberry committed:
disable run button on new workflow

davelopez commented on client/src/components/Workflow/List/useWorkflowActions.ts:
```diff
@@ -0,0 +1,110 @@
+import { getFullAppUrl } from "@/utils/utils";
+
+// TODO: replace me with a more accurate type
+type Workflow = any;
```
I'm not sure if the model you are looking for is already in:

https://github.com/galaxyproject/galaxy/blob/ff9470e1fc3d133416d2851fcdd36a87472b082f/client/src/api/workflows.ts#L3

But if it is not, it would be a good idea to add the type definition there and import the same definitions around (workflows.services.ts, WorkflowIndicators.vue, etc.) so when we have the proper model we just replace it in one place.

itisAliRH replied:
@davelopez Thanks for pointing it out! I addressed it in 9a22b0e853c64c07b37194fbc97fd8a8453eb2a8.

davelopez reviewed (commented):
I have played around with this and I love it!

Amazing work!

The tests would require some massaging to make them happy though... :sweat_smile:

itisAliRH commented on client/src/components/ActivityBar/ActivityBar.vue:
```diff
@@ -50,30 +95,30 @@ const dragItem: Ref<Activity | null> = ref(null);
+    return (
+        isActiveSideBar(activity.id) ||
+        (activity.to !== undefined && activity.to !== null && isActiveRoute(activity.to))
+    );
```
The `isActiveRoute` can handles `null/undefined`, so this line can be simpler:

```suggestion
    return (
        isActiveSideBar(activity.id) || isActiveRoute(activity.to)
    );
```

itisAliRH committed:
Updates bookmark toggle function parameters

Fixing bookmarking/unbookmarking workflows. Modifies the `toggleBookmark` function calls to include boolean arguments indicating the bookmark state.

itisAliRH committed:
Removes non-existent config property usage

Eliminates references to a non-existent config property *stored_workflow_menu_entries*  in useWorkflowActions to prevent errors and clean up the code. Client does not have access to this property anymore and is not used anywhere

itisAliRH committed:
Update client/src/components/ActivityBar/ActivityBar.vue

Remove redundant checks

itisAliRH committed:
Adds workflowId to useStores return object

Includes workflowId in the return object of useStores function in terminals unit test. Fixing test error.

itisAliRH committed:
Adds computed property for tools favorites toggle

Introduces a computed property to manage the favorites toggle state
Replaces the onFavorites function with a watcher for query changes
Simplifies the FavoritesButton binding using v-model.

itisAliRH committed:
Fixes FontAwesome icon usage in ToolPanel

Removes unnecessary library import and addition
Updates FontAwesomeIcon component to use dynamic icon binding

itisAliRH committed:
Fix the test to check the shared indicator tooltip text

itisAliRH committed:
Adds Pinia testing support to ActivityItem tests and add required prop activityBarId

itisAliRH committed:
Updates workflow type to StoredWorkflowDetailed

Replaces 'any' type with 'StoredWorkflowDetailed' for workflow props
Improves type safety and code readability in Workflow components

itisAliRH committed:
Fixes import path and adds activity bar ID to NotificationItem

Corrects the import path for ActivityItem component
Adds activity-bar-id required prop to NotificationItem

itisAliRH committed:
Adds activity-bar-id prop to ActivityItem components

Updates import path for ActivityItem in InteractiveItem.vue
Adds activity-bar-id prop to ActivityItem in InteractiveItem.vue and UploadItem.vue

itisAliRH committed:
Adds favorites filter functionality to WorkflowPanel

Integrates FavoritesButton component into WorkflowPanel
Implements computed property to manage bookmarked filter state
Updates filter text handling to support bookmarked filtering

itisAliRH committed:
Refactors icon imports and usage

Updates icon imports to use `IconDefinition` type from `@fortawesome/fontawesome-svg-core`
Replaces string-based icon references with FontAwesome icon objects
Removes unnecessary `library.add` calls

Improves type safety and consistency in icon usage in activity bar components

itisAliRH committed:
Updates icon handling in tests

Replaces stubbed icon component with FontAwesomeIcon in ActivitySettings test
Changes icon property type to IconDefinition in activityStore test for type safety

itisAliRH reviewed (approved):
@ElectronicBlueberry Great work, Laila!

I’ve played and reviewed it and have addressed some issues and also made some improvements.

Thank you so much!

itisAliRH committed:
Updates filter keyword for workflow bookmarks

Fix the filter keyword from "#favorites" to "is:bookmarked"

itisAliRH committed:
Simplifies activity item IDs

Removes redundant prefix from activity item IDs for consistency
Changes notification item ID to 'notifications' for clarity

itisAliRH committed:
Fix activity item upload progress background

itisAliRH committed:
Adds ID to toolbox panel

Adds an ID attribute to the toolbox panel.
Updates the tool opening logic in the workflow editor to open the toolbox panel using the activity bar before selecting tool

itisAliRH committed:
Adds auto-layout button ID in workflow toolbar

Introduces an ID for the auto-layout button in the toolbar component.
Updates navigation configuration and Selenium tests to use the new button ID for auto-layout functionality.

guerler commented on client/src/components/Markdown/MarkdownEditor.vue:
```diff
@@ -1,8 +1,5 @@
 <template>
     <div id="columns" class="d-flex">
-        <FlexPanel side="left">
-            <MarkdownToolBox :steps="steps" @onInsert="onInsert" />
```
Might need to be added back in the context of pages.

itisAliRH committed:
Adds hidePanel prop to ActivityBar and integrates MarkdownToolBox in MarkdownEditor

Introduces hidePanel prop to conditionally hide the side panel in ActivityBar.
Integrates MarkdownToolBox into MarkdownEditor temporary to have MarkdownToolBox in the page editor too.

ElectronicBlueberry committed:
remove wait for absent element

ElectronicBlueberry committed:
fix add_input

ElectronicBlueberry committed:
fix inserting steps

ElectronicBlueberry committed:
fix tool menu container occurrences

itisAliRH committed:
Migrates WorkflowAttributes test to TypeScript and updates dependencies

Renames WorkflowAttributes.test.js to WorkflowAttributes.test.ts
Replaces createLocalVue with getLocalVue and adds Pinia setup
Mocks userTagsStore methods and updates test assertions

ElectronicBlueberry committed:
fix workflows activity id

itisAliRH committed:
Migrates ToolPanel test to TypeScript

Updates imports to use TypeScript-compatible paths and extensions
Adjusts mock configuration and type assertions for TypeScript

itisAliRH committed:
Updates toolbox panel selectors and workflow editor logic

Renames toolbox panel selector for consistency
Adds new selector for workflow editor tools activity
Modifies logic to handle toolbox panel visibility and search functionality

itisAliRH committed:
Adds 'Save As' functionality to workflow editor

itisAliRH committed:
Adds 'Save As' option to workflow editor toolbar

Includes a new selector for the 'Save As' option in the navigation configuration.
Updates the Selenium test to handle the 'Save As' option visibility and interaction.

Improves user experience by ensuring the 'Save As' option is accessible and testable.

ElectronicBlueberry committed:
fix basics test

ElectronicBlueberry committed:
revert shortening of ids

ElectronicBlueberry committed:
fix test save as

ElectronicBlueberry committed:
add more wait time for search

ElectronicBlueberry committed:
re-add confirm dialog

ElectronicBlueberry committed:
remove double id prepending

ElectronicBlueberry committed:
redo add subworkflow

ElectronicBlueberry committed:
fix save as selector

ElectronicBlueberry committed:
move selection options to the left, so they do not interfere with the node inspector

ElectronicBlueberry committed:
remove collapse buttons from basics test

ElectronicBlueberry committed:
wait for save to complete

ElectronicBlueberry committed:
remove test for removed feature

ElectronicBlueberry committed:
remove invalid check in insert steps

ElectronicBlueberry committed:
remove test for removed component

ElectronicBlueberry committed:
make search selector more specific

ElectronicBlueberry committed:
fix selection test

ElectronicBlueberry committed:
clear search before searching for other tool

ElectronicBlueberry committed:
use confirm dialog composable

ElectronicBlueberry committed:
exclude node buttons from node selection targets

ElectronicBlueberry committed:
use latest id on insert

ElectronicBlueberry committed:
save latest id to dict

ElectronicBlueberry committed:
fix test_non_data_connections test

ElectronicBlueberry committed:
fix test_editor_duplicate_node test

davelopez committed:
fix event name for markdown update in PageEditorMarkdown component

After refactoring `move MarkdownEditor to activity bar`
